File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/part/PartComponent.ts
module api.content.page.part {

    import Region = api.content.page.region.Region;
    import PropertyTree = api.data.PropertyTree;
    import PropertyIdProvider = api.data.PropertyIdProvider;

    export class PartComponent extends api.content.page.DescriptorBasedComponent implements api.Equitable, api.Cloneable {

        constructor(builder: PartComponentBuilder) {
            super(builder);
        }

        toJson(): api.content.page.ComponentTypeWrapperJson {
            var json: PartComponentJson = <PartComponentJson>super.toPageComponentJson();

            return <api.content.page.ComponentTypeWrapperJson> {
                PartComponent: json
            };
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, PartComponent)) {
                return false;
            }

            return super.equals(o);
        }

        clone(generateNewPropertyIds: boolean = false): PartComponent {
            return new PartComponentBuilder(this, generateNewPropertyIds).build();
        }
    }

    export class PartComponentBuilder extends api.content.page.DescriptorBasedPageComponentBuilder<PartComponent> {

        constructor(source?: PartComponent, generateNewPropertyIds: boolean = false) {

            super(source, generateNewPropertyIds);
        }

        public fromJson(json: PartComponentJson, region: Region, propertyIdProvider: PropertyIdProvider): PartComponentBuilder {

            if (json.descriptor) {
                this.setDescriptor(api.content.page.DescriptorKey.fromString(json.descriptor));
            }
            this.setName(new api.content.page.ComponentName(json.name));
            this.setConfig(PropertyTree.fromJson(json.config, propertyIdProvider));
            this.setParent(region);
            return this;
        }

        public build(): PartComponent {
            return new PartComponent(this);
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/xp/core/content/page/region/PartComponent.java
package com.enonic.xp.core.content.page.region;


import com.enonic.xp.core.content.page.DescriptorKey;
import com.enonic.xp.core.data.PropertyTree;

public final class PartComponent
    extends DescriptorBasedComponent
{
    public PartComponent( final Builder builder )
    {
        super( builder );
    }

    public static Builder newPartComponent()
    {
        return new Builder();
    }

    public static Builder newPartComponent( final PartComponent source )
    {
        return new Builder( source );
    }

    public Component copy()
    {
        return newPartComponent( this ).build();
    }

    @Override
    public ComponentType getType()
    {
        return PartComponentType.INSTANCE;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof PartComponent ) )
        {
            return false;
        }

        return super.equals( o );
    }

    public static class Builder
        extends DescriptorBasedComponent.Builder
    {
        private Builder()
        {
            // Default
        }

        private Builder( final PartComponent source )
        {
            super( source );
        }

        public Builder name( ComponentName value )
        {
            this.name = value;
            return this;
        }

        public Builder name( String value )
        {
            this.name = value != null ? new ComponentName( value ) : null;
            return this;
        }

        public Builder descriptor( DescriptorKey value )
        {
            this.descrpitor = value;
            return this;
        }

        public Builder descriptor( String value )
        {
            this.descrpitor = DescriptorKey.from( value );
            return this;
        }

        public Builder config( final PropertyTree config )
        {
            this.config = config;
            return this;
        }

        public PartComponent build()
        {
            return new PartComponent( this );
        }
    }
}
